js/src/ipc/writer/binary.ts

```diff
@@ -0,0 +1,705 @@
+    const dataOffset = writeRecordBatch(b, node.data);
+    return (
+        _DictionaryBatch.startDictionaryBatch(b) ||
+        _DictionaryBatch.addId(b, new Long(node.id, 0)) ||
+        _DictionaryBatch.addIsDelta(b, node.isDelta) ||
+        _DictionaryBatch.addData(b, dataOffset) ||
+        _DictionaryBatch.endDictionaryBatch(b)
+    );
+}
+
+function writeBuffer(b: Builder, node: BufferMetadata) {
+    return _Buffer.createBuffer(b, new Long(node.offset, 0), new Long(node.length, 0));
+}
+
+function writeFieldNode(b: Builder, node: FieldMetadata) {
+    return _FieldNode.createFieldNode(b, new Long(node.length, 0), new Long(node.nullCount, 0));
+}
+
+function writeField(b: Builder, node: Field) {
+    let typeOffset = -1;
+    let type = node.type;
+    let typeId = node.typeId;
+    let name: number | undefined = undefined;
+    let metadata: number | undefined = undefined;
+    let dictionary: number | undefined = undefined;
+
+    if (!DataType.isDictionary(type)) {
+        typeOffset = new TypeSerializer(b).visit(type);
+    } else {
+        typeId = type.dictionary.TType;
+        dictionary = new TypeSerializer(b).visit(type);
+        typeOffset = new TypeSerializer(b).visit(type.dictionary);
+    }
+
+    let children = _Field.createChildrenVector(b, (type.children || []).map((f) => writeField(b, f)));
+    if (node.metadata && node.metadata.size > 0) {
+        metadata = _Field.createCustomMetadataVector(
+            b,
+            [...node.metadata].map(([k, v]) => {
+                const key = b.createString(k);
+                const val = b.createString(v);
+                return (
+                    _KeyValue.startKeyValue(b) ||
+                    _KeyValue.addKey(b, key) ||
+                    _KeyValue.addValue(b, val) ||
+                    _KeyValue.endKeyValue(b)
+                );
+            })
+        );
+    }
+    if (node.name) {
+        name = b.createString(node.name);
+    }
+    return (
+        _Field.startField(b) ||
+        _Field.addType(b, typeOffset) ||
+        _Field.addTypeType(b, typeId) ||
+        _Field.addChildren(b, children) ||
+        _Field.addNullable(b, !!node.nullable) ||
+        (name !== undefined && _Field.addName(b, name)) ||
+        (dictionary !== undefined && _Field.addDictionary(b, dictionary)) ||
+        (metadata !== undefined && _Field.addCustomMetadata(b, metadata)) ||
+        _Field.endField(b)
+    );
+}
+
+function mapReverse<T, U>(source: T[], callbackfn: (value: T, index: number, array: T[]) => U): U[] {
+    const result = new Array(source.length);
+    for (let i = -1, j = source.length; --j > -1;) {
+        result[i] = callbackfn(source[j], i, source);
+    }
+    return result;
+}
+
+const platformIsLittleEndian = (function() {
+    const buffer = new ArrayBuffer(2);
+    new DataView(buffer).setInt16(0, 256, true /* littleEndian */);
+    // Int16Array uses the platform's endianness.
+    return new Int16Array(buffer)[0] === 256;
```

Comment:
This is an interesting tidbit, is there a reference on this?

Reply:
Oh, I forgot I dropped that in there -- should probably be put into one of our util functions. This little helper is from the [MDN example](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) for the JS DataView class.